### PR TITLE
fix: include interfile dependencies in Snyk Code scans [IDE-182] [IDE-178]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
https://snyksec.atlassian.net/browse/IDE-178

### Description

_Provide description of this PR and changes, if linked Jira ticket doesn't cover it in full._

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [ ] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

🚨After having merged, please update the CLI go.mod to pull in latest language server.

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: domain/ide/workspace/folder.go [hunk 1]
```diff
@@ -77,6 +79,28 @@ type Folder struct {
 	notifier                noti.Notifier
 }
 
+func (f *Folder) Issues() map[string][]snyk.Issue {
+	issues := make(map[string][]snyk.Issue)
+	f.documentDiagnosticCache.Range(func(key string, value []snyk.Issue) bool {
```
Comment (bastiandoetsch) [inline domain/ide/workspace/folder.go:84]: let's add some comments regarding global/local caching

File: internal/util/identity_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+package util
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_getIssueId(t *testing.T) {
```
Comment (teodora-sandu) [inline internal/util/identity_test.go:25]:
```suggestion
func Test_GetIssueKey(t *testing.T) {
```

File: infrastructure/code/code.go [hunk 3]
```diff
@@ -166,43 +175,80 @@ func (sc *Scanner) Scan(ctx context.Context, path string, folderPath string) (is
...
 
-func (sc *Scanner) waitForScanToFinish(scanStatus *ScanStatus, folderPath string) (waiting bool) {
+// getFilesToBeScanned returns a map of files that need to be scanned and removes them from the changedPaths set.
+// This function also analyzes interfile dependencies, taking into account the dataflow between files.
+func (sc *Scanner) getFilesToBeScanned(folderPath string) map[string]bool {
+	logger := config.CurrentConfig().Logger().With().Str("method", "code.getFilesToBeScanned").Logger()
+	changedFiles := make(map[string]bool)
+	for changedPath := range sc.changedPaths[folderPath] {
+		if uri.IsDirectory(changedPath) {
+			logger.Debug().Str("path", changedPath).Msg("skipping directory")
```
Comment (teodora-sandu) [inline infrastructure/code/code.go:212]: I'll need to do a manual test but I noticed in the old code we would run `delete(sc.changedPaths[folderPath], changedPath)` regardless. I wonder if this makes a difference to anything

Comment (teodora-sandu) [reply infrastructure/code/code.go:212]: I don't think it causes any real side effects actually. The only thing I could see happening is if there are multiple scans queued at once and an earlier scan goes through the `changedPaths`, the next scan would still need to see that the left over paths in `changedPaths` are directories and skip them.

File: infrastructure/code/code_test.go [hunk 4]
```diff
@@ -892,3 +896,159 @@ func Test_SastApiCall(t *testing.T) {
...
+
+func TestScanner_getFilesToBeScanned(t *testing.T) {
+	config.CurrentConfig().SetSnykCodeEnabled(true)
+	_, scanner := setupTestScanner(t)
+	tempDir := t.TempDir()
+	scanner.changedPaths = make(map[string]map[string]bool)
+	scanner.changedPaths[tempDir] = make(map[string]bool)
+
+	t.Run("should add all files from changedPaths map and delete them from changedPaths", func(t *testing.T) {
+		changedFile := "file1.java"
```
Comment (teodora-sandu) [inline infrastructure/code/code_test.go:908]: Not too sure why this is stored in a variable and the other isn't?

File: infrastructure/code/code_test.go [hunk 5]
```diff
@@ -892,3 +896,159 @@ func Test_SastApiCall(t *testing.T) {
...
+		require.Contains(t, files, "file2.java")
+		require.Len(t, scanner.changedPaths[tempDir], 0)
+	})
+
+	t.Run("should add all files that have dataflow items of a changed file", func(t *testing.T) {
+		changedFile := "main.ts"
+		fromChangeAffectedFile := "juice-shop/routes/vulnCodeSnippet.ts"
+
+		// we need to add a fake issue to the cache, let's keep it to the minimum needed
+		// we are reusing the test data from the code_html test
```
Comment (teodora-sandu) [inline infrastructure/code/code_test.go:924]: I can see us changing some of the logic in that code_html file or test and then we break this. Would it not be better and clearer to just hardcode it here? It'll also make it more obvious why this test does what it does

**Comment by teodora-sandu** [comment]
I tried to manually test it with a simple repo https://github.com/snyk/target-service and it failed 😕 

```
2024-04-11 17:35:55,018 [  12838]   WARN - #io.snyk.plugin.ui.SnykBalloonNotifications - scanning error for project target-service, emptying cache.Data: SnykScanParams(status=error, product=code, folderPath=/Users/teodorasandu/Documents/repos/target-service, issues=null)
```

Screenshot 2024-04-11 at 17 34 41

Not sure if I've done something wrong or I've found an edge case by accident.

**Review by teodora-sandu — COMMENTED** [review]
I was just about to approve it when the manual test I ran errored. I'm not sure if it's my setup or not. I won't block it if you want to get another opinion, since I'm off tomorrow 🙇

File: infrastructure/code/code_test.go (hunk 5 above)
Comment (bastiandoetsch) [reply infrastructure/code/code_test.go:924]: Yeah, I was lazy ;)

**Comment by bastiandoetsch** [comment]
I'll try it out tomorrow, didn't see the issue though. Maybe something related to the new flow - we'll see.

**Comment by bastiandoetsch** [comment]
> I tried to manually test it with a simple repo https://github.com/snyk/target-service and it failed 😕
> 
> ```
> 2024-04-11 17:35:55,018 [  12838]   WARN - #io.snyk.plugin.ui.SnykBalloonNotifications - scanning error for project target-service, emptying cache.Data: SnykScanParams(status=error, product=code, folderPath=/Users/teodorasandu/Documents/repos/target-service, issues=null)
> ```
> 
> Screenshot 2024-04-11 at 17 34 41
> 
> Not sure if I've done something wrong or I've found an edge case by accident.
> 
> This is my setup: Screenshot 2024-04-11 at 17 38 41

I think it is caused by the workspace API not ready yet in prod. 
```
2024-04-12 08:32:54,065 [ 486578]   INFO - Snyk Language Server - 2024-04-12T08:32:54+02:00 INF - creating workspace bundleHash=xxx path=/Users/bdoetsch/workspace/juice-shop repositoryUri=https://github.com/juice-shop/juice-shop.git

2024-04-12 08:32:54,238 [ 486751]   INFO - Snyk Language Server - 2024-04-12T08:32:54+02:00 ERR processResults - Product returned an error error= product="Snyk Code"
```

File: domain/snyk/scanner.go [hunk 6]
```diff
@@ -241,3 +254,46 @@ func getEnabledAnalysisTypes(productScanners []ProductScanner) (analysisTypes []
...
+		}
+	}
+	return issues
+}
+
+func (sc *DelegatingConcurrentScanner) Issue(key string) Issue {
+	for _, scanner := range sc.scanners {
+		if s, ok := scanner.(IssueProvider); ok {
+			issue := s.Issue(key)
+			if issue.ID != "" {
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply domain/snyk/scanner.go:272]: 0-values in golang are tricky to compare, and if we don't pass a pointer, we cannot compare to nil. So I'm using the ID field as substitute for checking if the issue is empty. We could use title or key, too - it's not really important which of the mandatory fields are checked.

File: infrastructure/code/code.go (hunk 3 above)
Comment (bastiandoetsch) [reply infrastructure/code/code.go:212]: Yeah, it should not make a difference. Also, it's cleaner to have the guard first in the loop, I think.

**Review by teodora-sandu — APPROVED** [review]
